File: 139_Word_Break/139_2.py
from typing import List

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        wordSet = set(wordDict)
        size = len(s)
        import functools
        @functools.lru_cache(None)
        def dfs(start) -> bool:
            if start >= size:
                return True
            for j in range(start + 1, size + 1):
                if s[start: j] in wordSet and dfs(j):
                    return True
            return False
        return dfs(0)

File: 139_Word_Break/139_2.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int size = s.length();
        Set<String> wordSet = new HashSet<>(wordDict);
        Map<Integer, Boolean> memo = new HashMap<>();
        return dfs(0, s, size, wordSet, memo);
    }

    private boolean dfs(int i, String s, int size, Set<String> wordSet, Map<Integer, Boolean> memo) {
        if (i >= size) return true;
        if (memo.containsKey(i)) return memo.get(i);
        for (int j = i + 1; j < size + 1; j++) {
            if (wordSet.contains(s.substring(i, j)) && dfs(j, s, size, wordSet, memo)) {
                memo.put(i, true);
                return true;
            }
        }
        memo.put(i, false);
        return false;
    }
}
